### DISCUSSION TITLE
30 Nov 2022 - ILP Summit Catch Up

### DISCUSSION THREAD

**Discussion by sabineschaller** [interledger/roadmap#28] [Ideas]
Here is a list of things we want to discuss today including links to resources:
* the auth server as a policy agent ([Be more GNAP like](https://www.notion.so/coil/Be-more-GNAP-like-8dffb1620ffd4f16bddc9f2911b9dec8))
* Rafiki without STREAM
  * https://github.com/interledger/rafiki/issues/747
* Alternatives to STREAM in Open Payments
  *  https://github.com/interledger/open-payments/issues/219
* Replace Open Payments `description` and `externalRef` with `metadata` or `memos`
  * https://github.com/interledger/open-payments/issues/221

**Comment by wilsonianb** [comment]
This seems related to policy agent / more GNAP-like:
- https://github.com/interledger/open-payments/issues/206

**Comment by wilsonianb** [comment]
>     * Rafiki without STREAM
>       
>       * [Decouple STREAM rafiki#747](https://github.com/interledger/rafiki/issues/747)
> 
>     * Alternatives to STREAM in Open Payments
>       
>       * [Replace `ilpStreamConnection` with methods/transports open-payments#219](https://github.com/interledger/open-payments/issues/219)

IMO, these are a single topic that falls under that broader topic of Interledger the protocol vs Interledger the idea/mission:
- https://github.com/interledger/open-payments/issues/219#issuecomment-1326577546

**Comment by adrianhopebailie** [comment]
To expand on the first idea:

- Payment Pointers should point to the AS (not an RS)
- The AS should simply be an implementation of the GNAP protocol (although maybe still a limited profile is REQUIRED) so there is very little extra work to do for us in the definition of the AS interface
- We need to define standard **grant schemas** and the semantics of what they mean
- The grant schemas SHOULD NOT map 1:1 onto resources at the RS they SHOULD map to "use cases" and are therefor easier to represent to users for consent. E.g. "Payment of $5 to Bob from default checking account 1234"
- A grant could be usable at different RS instances and types (e.g. Open Banking, Open Payments, PayPal etc)
- The AS should provide the locations of RS instances where the grant can be used in the grant response

This model aligns better with GNAP and the AS/RS separation. The RS can still apply policies to requests that come in with a grant token (i.e. the grant allows for X but the RS independently denies X due to internal policy (tx limit or similar))

This does mean we need a way to signal to the RS how/where it can introspect a token. GNAP has some pretty good guidance on this and I think a lot of what @wilsonianb and @sabineschaller have done is already aligned with this spec: https://www.ietf.org/archive/id/draft-ietf-gnap-resource-servers-02.html

This would also make Open Payments a purely RS API (I think) and would make it much cleaner as we can now pass the payment pointer for a resource INSIDE the requests and have a single URL on a server for all resource requests (for all payment pointers).

**Reply by wilsonianb** [reply]
Would the AS translate a grant schema (in the grant request) to specific resource collection(s) that have been registered by the relevant RS?
- https://github.com/interledger/open-payments/issues/206#issuecomment-1290650584

I'm particularly looking at the "Example 2: Web Monetization" in the notion doc, in which the grant response doesn't match the grant schema specified in the request.

**Reply by wilsonianb** [reply]
>we can now pass the payment pointer for a resource INSIDE the requests

I'm trying to grasp the AS's relationship with the RO vs the RS.
It sounding more like the RO having a relationship with the AS (that is where their payment pointer lives).
But then it seems like the RS would also need to register its resource set at the AS in order for the resource endpoints to be included in the grant response.
And if the payment pointer is an AS thing, why would it be included in resource requests, as opposed to maybe the RO's RS specific details being included in the grant response? Or is it that the RS would query the payment to determine the relevant RO?
